src/EnergyPlus/DaylightingDevices.cc

```diff
@@ -1620,6 +1614,164 @@ namespace DaylightingDevices {
+            ShowContinueError(
+                state,
+                "and possibly also the view factor to the ground. Check you input and/or consider turning off autosizing of the view factors.");
+            leftoverViewFactor = 1.0 - viewFactorToShelf - viewFactorToGround;
+            if (leftoverViewFactor >= 0.0) {
+                viewFactorToSky = leftoverViewFactor;
+            } else {
+                viewFactorToSky = 0.0;
+                viewFactorToGround = 1.0 - viewFactorToShelf;
+                if (viewFactorToGround < 0.0) {
+                    viewFactorToGround = 0.0;
+                    viewFactorToShelf = 1.0;
+                }
+            }
+        } else { // At least part of the shelf is somewhere in the middle of the window so we need to split out the view factors
+            ShowContinueError(
+                state,
+                "Since the light shelf is neither fully above or fully below the window to which it is associated, the view factor of the window");
+            ShowContinueError(
+                state,
+                "to the ground and sky were both potentially reduced Check you input and/or consider turning off autosizing of the view factors.");
+            Real64 zShelfAvg;
+            if (((zShelfMin >= zWinMin) && (zShelfMax <= zWinMax)) || // Shelf does not go above or below the window
+                ((zShelfMin < zWinMin) && (zShelfMax > zWinMax))) {   // Shelf goes both above AND below the window
+                zShelfAvg = 0.5 * (zShelfMin + zShelfMax);
+            } else if (zShelfMin < zWinMin) { // Shelf goes partially below the window only
+                Real64 fracAbove = 0.0;
+                if (zShelfMax > zShelfMin) {
+                    fracAbove = (zShelfMax - zWinMin) / (zShelfMax - zShelfMin);
+                    if (fracAbove > 1.0) fracAbove = 1.0;
+                }
+                zShelfAvg = zWinMin + fracAbove * (zShelfMax - zWinMin);
+            } else { // (zShelfMax > zWinMax): Shelf goes partially above window
+                Real64 fracBelow = 0.0;
+                if (zShelfMax > zShelfMin) {
+                    fracBelow = (zWinMax - zShelfMin) / (zShelfMax - zShelfMin);
+                }
+                zShelfAvg = zWinMax - fracBelow * (zWinMax - zShelfMin);
+            }
+
+            // Find height ratio based on shelf average height
+            Real64 heightRatio;
+            if (zWinMax > zWinMin) { // Window has a positive height
+                heightRatio = (zShelfAvg - zWinMin) / (zWinMax - zWinMin);
+                heightRatio = min(heightRatio, 1.0);
+                heightRatio = max(heightRatio, 0.0);
+            } else { // Window does not have a positive height (not realistic) so set height ratio based on shelf location
+                if (zShelfAvg > zWinMax) {
+                    heightRatio = 1.0;
```

Comment:
@jcyuan2020 I'm not sure I understand your comment here.  Starting in line 1742, we have to have some window height to avoid a divide by zero in heightRatio.  The else that starts at 1746 should never happen but I'm just making sure the case is covered.  Are you saying lines 1747 through 1751 should just be replaced by heightRatio = 1.0?  That just checks where the shelf is with respect to the window.  I guess it's not necessary, but I want to make sure I understand what you are saying correctly so I get this right.

Reply:
Sorry @RKStrand , my bad! I thought the line in the `else` bracket was also `=1.0` when browsing through it initially.

Reply:
No problem--the code here is a little tedious and it's easy to miss that detail.